sdk/azcore/arm/resource_identifier.go

```diff
@@ -0,0 +1,315 @@
+	subscriptionId    string
+	provider          string
+	resourceGroupName string
+	location          string
+	resourceType      ResourceType
+	name              string
+	isChild           bool
+
+	stringValue *string
+}
+
+func (id ResourceIdentifier) Parent() *ResourceIdentifier {
+	return id.parent
+}
+
+func (id ResourceIdentifier) SubscriptionId() string {
+	return id.subscriptionId
+}
+
+func (id ResourceIdentifier) Provider() string {
+	return id.provider
+}
+
+func (id ResourceIdentifier) ResourceGroupName() string {
+	return id.resourceGroupName
+}
+
+func (id ResourceIdentifier) Location() string {
+	return id.location
+}
+
+func (id ResourceIdentifier) ResourceType() ResourceType {
+	return id.resourceType
+}
+
+func (id ResourceIdentifier) Name() string {
+	return id.name
+}
+
+func newResourceIdentifier(parent *ResourceIdentifier, resourceTypeName string, resourceName string) *ResourceIdentifier {
+	id := &ResourceIdentifier{}
+	id.init(parent, chooseResourceType(resourceTypeName, parent), resourceName, true)
+	return id
+}
+
+func newResourceIdentifierWithResourceType(parent *ResourceIdentifier, resourceType ResourceType, resourceName string) *ResourceIdentifier {
+	id := &ResourceIdentifier{}
+	id.init(parent, resourceType, resourceName, true)
+	return id
+}
+
+func newResourceIdentifierWithProvider(parent *ResourceIdentifier, providerNamespace, resourceTypeName, resourceName string) *ResourceIdentifier {
+	id := &ResourceIdentifier{}
+	id.init(parent, NewResourceType(providerNamespace, resourceTypeName), resourceName, false)
+	return id
+}
+
+func chooseResourceType(resourceTypeName string, parent *ResourceIdentifier) ResourceType {
+	switch strings.ToLower(resourceTypeName) {
+	case resourceGroupsLowerKey:
+		return ResourceGroupResourceType
+	case subscriptionsKey:
+		return SubscriptionResourceType
+	default:
+		return NewResourceTypeFromParent(parent.resourceType, resourceTypeName)
+	}
+}
+
+func (id *ResourceIdentifier) init(parent *ResourceIdentifier, resourceType ResourceType, name string, isChild bool) {
+	if parent != nil {
+		id.provider = parent.provider
+		id.subscriptionId = parent.subscriptionId
+		id.resourceGroupName = parent.resourceGroupName
+		id.location = parent.location
+	}
+
+	if resourceType.String() == SubscriptionResourceType.String() {
+		id.subscriptionId = name
+	}
+
+	if resourceType.LastType() == locationsKey {
+		id.location = name
+	}
+
+	if resourceType.String() == ResourceGroupResourceType.String() {
+		id.resourceGroupName = name
+	}
+
+	if resourceType.String() == ProvidersResourceType.String() {
+		id.provider = name
+	}
+
+	if parent == nil {
+		id.parent = RootResourceIdentifier
+	} else {
+		id.parent = parent
+	}
+	id.isChild = isChild
+	id.resourceType = resourceType
+	id.name = name
+}
+
+// ParseResourceIdentifier parses a string to an instance of ResourceIdentifier
+func ParseResourceIdentifier(id string) (*ResourceIdentifier, error) {
+	if len(id) == 0 {
```

Comment:
Seems like we could use a regex to validate the input and pull out the interesting parts.